WebRTC MediaCapabilitiesDecodingInfo.configuration should not be undefined if supported is false

youennf opened the pull request:
#### 2f4c13b7dde46e6dd0df0a6e147e6147eed6768e
<pre>
WebRTC MediaCapabilitiesDecodingInfo.configuration should not be undefined if supported is false
<a href="https://rdar.apple.com/150680756">rdar://150680756</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=292541">https://bugs.webkit.org/show_bug.cgi?id=292541</a>

Reviewed by Jean-Yves Avenard.

Make sure to return a non default configuration in the non supported WebRTC code path, as per specification.
We also update the dictionaries web idl to mark configuration as required to make it clear it should not be optional.

Covered by updated test.

* LayoutTests/webrtc/vp9-tests.js:
(promise_test.async if):
(promise_test):
(promise_test.async let):
(promise_test.async test):
(promise_test.then.async video):
(test): Deleted.
(hasVP9.promise_test.async let): Deleted.
(hasVP9.promise_test): Deleted.
(hasVP9.promise_test.async test): Deleted.
(hasVP9.promise_test.then.async video): Deleted.
* Source/WebCore/Modules/mediacapabilities/MediaCapabilitiesDecodingInfo.idl:
* Source/WebCore/Modules/mediacapabilities/MediaCapabilitiesEncodingInfo.idl:
* Source/WebCore/platform/mediastream/WebRTCProvider.cpp:
(WebCore::WebRTCProvider::createDecodingConfiguration):
(WebCore::WebRTCProvider::createEncodingConfiguration):

Canonical link: <a href="https://commits.webkit.org/294957@main">https://commits.webkit.org/294957@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/92d7a2da052ba37566c42308471f50b52c31d8f2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103616 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23299 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13619 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108792 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54263 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/105656 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23654 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31849 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78736 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106622 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18350 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93472 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59071 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/18161 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11519 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53617 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/87944 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11578 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111170 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30763 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22686 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/87729 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31124 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89672 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87376 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22243 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32255 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9976 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25117 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30691 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/36003 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30491 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33822 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32052 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jernoble reviewed (changes requested):
I don't think this is a correct reading of the spec, and even if it is, the spec should be changed to return a valid `supportedConfiguration`/`configuration` property in the output. The point of `configuration` is to allow feature detection of supported or unsupported properties in the `VideoDecodingInfo` and `AudioDecodingInfo` dictionaries. That use case is still valid if `supported === false`, if not more so.

jyavenard commented on LayoutTests/webrtc/vp9-tests.js:
```diff
@@ -30,14 +34,18 @@ test(() => {
+    const videoConfiguration = { contentType: 'video/vp9', width: 800, height: 600, bitrate: 3000, framerate: 24 };
+    const results = await navigator.mediaCapabilities.decodingInfo({type: 'webrtc', video: videoConfiguration });
+    assert_false(results.supported, "decoder MC supported");
+    assert_not_equals(results.supportedConfiguration, undefined, "decoder MC configuration");
```
supportedConfiguration isn't per spec anymore. We should be amending this

youennf replied:
I'll rebase the PR now that renaming PR landed.

jyavenard commented on LayoutTests/webrtc/vp9-tests.js:
```diff
@@ -1,8 +1,12 @@
 
+    const videoConfiguration = { contentType: 'video/vp9', width: 800, height: 600, bitrate: 3000, framerate: 24 };
+    const results = await navigator.mediaCapabilities.encodingInfo({type: 'webrtc', video: videoConfiguration });
+    assert_false(results.supported, "decoder MC supported");
```
s/decoder MC supported/decoder MC should not be supported/

jyavenard commented on LayoutTests/webrtc/vp9-tests.js:
```diff
@@ -30,14 +34,18 @@ test(() => {
 
+    const videoConfiguration = { contentType: 'video/vp9', width: 800, height: 600, bitrate: 3000, framerate: 24 };
+    const results = await navigator.mediaCapabilities.decodingInfo({type: 'webrtc', video: videoConfiguration });
+    assert_false(results.supported, "decoder MC supported");
```
same here, it's confusing otherwise.

youennf committed:
WebRTC MediaCapabilitiesDecodingInfo.configuration should not be undefined if supported is false
rdar://150680756
https://bugs.webkit.org/show_bug.cgi?id=292541

Reviewed by Jean-Yves Avenard.

Make sure to return a non default configuration in the non supported WebRTC code path, as per specification.
We also update the dictionaries web idl to mark configuration as required to make it clear it should not be optional.

Covered by updated test.

* LayoutTests/webrtc/vp9-tests.js:
(promise_test.async if):
(promise_test):
(promise_test.async let):
(promise_test.async test):
(promise_test.then.async video):
(test): Deleted.
(hasVP9.promise_test.async let): Deleted.
(hasVP9.promise_test): Deleted.
(hasVP9.promise_test.async test): Deleted.
(hasVP9.promise_test.then.async video): Deleted.
* Source/WebCore/Modules/mediacapabilities/MediaCapabilitiesDecodingInfo.idl:
* Source/WebCore/Modules/mediacapabilities/MediaCapabilitiesEncodingInfo.idl:
* Source/WebCore/platform/mediastream/WebRTCProvider.cpp:
(WebCore::WebRTCProvider::createDecodingConfiguration):
(WebCore::WebRTCProvider::createEncodingConfiguration):

Canonical link: https://commits.webkit.org/294957@main
